go generate ./... broken on Go 1.12+

The repository uses `go generate ./...` to generate the required and assertion formatter files, but when following the instructions in the `README.md` and following the example in `.travisci.gogenerate.sh` to generate the files with either Go 1.12.13 or 1.13.4, the following errors occur:

```
2019/11/09 22:18:23 open /home/leighmcculloch/devel/testify/src/github.com/stretchr/testify/assert/assertion_format.go: no such file or directory
exit status 1
assert/assertions.go:25: running "go": exit status 1
2019/11/09 22:18:24 open /home/leighmcculloch/devel/testify/src/github.com/stretchr/testify/assert/assertion_format.go: no such file or directory
exit status 1
require/forward_requirements.go:16: running "go": exit status 1
```

It looks like this code in the generator that looks up where the code lives is finding the code path incorrectly in these newer versions of Go that support Go modules. The code is located on my system outside a GOPATH.

https://github.com/stretchr/testify/blob/e442a1631a73174a4afde5b4c41ba21cde6f6889/_codegen/main.go#L174-L184

If I try moving the project to my GOPATH, I get a new set of errors:
```
2019/11/09 22:23:48 assertions.go:20:2: could not import github.com/davecgh/go-spew/spew (can't find import: "github.com/stretchr/testify/vendor/github.com/davecgh/go-spew/spew")
exit status 1
assert/assertions.go:25: running "go": exit status 1
2019/11/09 22:23:48 assertions.go:20:2: could not import github.com/davecgh/go-spew/spew (can't find import: "github.com/stretchr/testify/vendor/github.com/davecgh/go-spew/spew")
exit status 1
require/forward_requirements.go:16: running "go": exit status 1
```

--- a/.travis.gogenerate.sh
+++ b/.travis.gogenerate.sh
@@ -3,7 +3,7 @@
 # If GOMOD is defined we are running with Go Modules enabled, either
 # automatically or via the GO111MODULE=on environment variable. Codegen only
 # works with modules, so skip generation if modules is not in use.
-if [[ ! -z "$(go env GOMOD)" ]]; then
+if [[ -z "$(go env GOMOD)" ]]; then
   exit 0
 fi
 

--- a/.travis.gogenerate.sh
+++ b/.travis.gogenerate.sh
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-go generate ./...
+GO111MODULE=on go generate ./...
 if [ -n "$(git diff)" ]; then
   echo "Go generate had not been run"
   git diff

--- a/.travis.gogenerate.sh
+++ b/.travis.gogenerate.sh
@@ -1,10 +1,13 @@
 #!/bin/bash
 
-if [[ "$TRAVIS_GO_VERSION" =~ ^1\.(12|13)(\..*)?$ ]]; then
+# If GOMOD is defined we are running with Go Modules enabled, either
+# automatically or via the GO111MODULE=on environment variable. Codegen only
+# works with modules, so skip generation if modules is not in use.
+if [[ ! -z "$(go env GOMOD)" ]]; then
   exit 0
 fi
 
-GO111MODULE=on go generate ./...
+go generate ./...
 if [ -n "$(git diff)" ]; then
   echo "Go generate had not been run"
   git diff

--- a/.travis.gogenerate.sh
+++ b/.travis.gogenerate.sh
@@ -1,10 +1,5 @@
 #!/bin/bash
 
-if [[ "$TRAVIS_GO_VERSION" =~ ^1\.[45](\..*)?$ ]]; then
-  exit 0
-fi
-
-go get github.com/ernesto-jimenez/gogen/imports
 go generate ./...
 if [ -n "$(git diff)" ]; then
   echo "Go generate had not been run"

--- a/.travis.gogenerate.sh
+++ b/.travis.gogenerate.sh
@@ -1,5 +1,9 @@
 #!/bin/bash
 
+if [[ "$TRAVIS_GO_VERSION" =~ ^1\.(12|13)(\..*)?$ ]]; then
+  exit 0
+fi
+
 GO111MODULE=on go generate ./...
 if [ -n "$(git diff)" ]; then
   echo "Go generate had not been run"

--- a/.travis.gogenerate.sh
+++ b/.travis.gogenerate.sh
@@ -4,6 +4,7 @@
 # automatically or via the GO111MODULE=on environment variable. Codegen only
 # works with modules, so skip generation if modules is not in use.
 if [[ -z "$(go env GOMOD)" ]]; then
+  echo "Skipping go generate because modules not enabled and required"
   exit 0
 fi
 

--- a/.travis.govet.sh
+++ b/.travis.govet.sh
@@ -1,10 +1,5 @@
 #!/bin/bash
 
-cd "$(dirname $0)"
-DIRS=". assert require mock _codegen"
 set -e
-for subdir in $DIRS; do
-  pushd $subdir
-  go vet
-  popd
-done
+
+go vet ./...

--- a/.travis.yml
+++ b/.travis.yml
@@ -11,6 +11,14 @@ matrix:
       env: GO111MODULE=off
     - go: "1.11.x"
       env: GO111MODULE=on
+    - go: "1.12.x"
+      env: GO111MODULE=off
+    - go: "1.12.x"
+      env: GO111MODULE=on
+    - go: "1.13.x"
+      env: GO111MODULE=off
+    - go: "1.13.x"
+      env: GO111MODULE=on
     - go: tip
   script:
     - ./.travis.gogenerate.sh

--- a/.travis.yml
+++ b/.travis.yml
@@ -20,8 +20,8 @@ matrix:
     - go: "1.13.x"
       env: GO111MODULE=on
     - go: tip
-script:
-  - ./.travis.gogenerate.sh
-  - ./.travis.gofmt.sh
-  - ./.travis.govet.sh
-  - go test -v -race $(go list ./... | grep -v vendor)
+  script:
+    - ./.travis.gogenerate.sh
+    - ./.travis.gofmt.sh
+    - ./.travis.govet.sh
+    - go test -v -race $(go list ./... | grep -v vendor)

--- a/.travis.yml
+++ b/.travis.yml
@@ -24,4 +24,4 @@ script:
   - ./.travis.gogenerate.sh
   - ./.travis.gofmt.sh
   - ./.travis.govet.sh
-  - go test -v -race $(go list ./... | grep -v vendor)
+  - go test -v -race ./...

--- a/.travis.yml
+++ b/.travis.yml
@@ -20,8 +20,8 @@ matrix:
     - go: "1.13.x"
       env: GO111MODULE=on
     - go: tip
-  script:
-    - ./.travis.gogenerate.sh
-    - ./.travis.gofmt.sh
-    - ./.travis.govet.sh
-    - go test -v -race $(go list ./... | grep -v vendor)
+script:
+  - ./.travis.gogenerate.sh
+  - ./.travis.gofmt.sh
+  - ./.travis.govet.sh
+  - go test -v -race $(go list ./... | grep -v vendor)

--- a/README.md
+++ b/README.md
@@ -321,7 +321,7 @@ To update Testify to the latest version, use `go get -u github.com/stretchr/test
 Supported go versions
 ==================
 
-We support the three major Go versions, which are 1.9, 1.10, and 1.11 at the moment.
+We support the three major Go versions, which are 1.11, 1.12, and 1.13 at the moment.
 
 ------
 

--- a/_codegen/.gitignore
+++ b/_codegen/.gitignore
@@ -1,0 +1,1 @@
+_codegen

--- a/_codegen/go.mod
+++ b/_codegen/go.mod
@@ -1,5 +1,5 @@
 module github.com/stretchr/testify/_codegen
 
-go 1.13
+go 1.11
 
 require github.com/ernesto-jimenez/gogen v0.0.0-20180125220232-d7d4131e6607

--- a/_codegen/go.mod
+++ b/_codegen/go.mod
@@ -1,0 +1,5 @@
+module github.com/stretchr/testify/_codegen
+
+go 1.13
+
+require github.com/ernesto-jimenez/gogen v0.0.0-20180125220232-d7d4131e6607

--- a/_codegen/main.go
+++ b/_codegen/main.go
@@ -194,7 +194,7 @@ func parsePackageSource(pkg string) (*types.Scope, *doc.Package, error) {
 	}
 
 	cfg := types.Config{
-		Importer: importer.ForCompiler(fset, "source", nil),
+		Importer: importer.For("source", nil),
 	}
 	info := types.Info{
 		Defs: make(map[*ast.Ident]types.Object),

--- a/_codegen/main.go
+++ b/_codegen/main.go
@@ -181,7 +181,7 @@ func parsePackageSource(pkg string) (*types.Scope, *doc.Package, error) {
 	files := make(map[string]*ast.File)
 	fileList := make([]*ast.File, len(pd.GoFiles))
 	for i, fname := range pd.GoFiles {
-		src, err := ioutil.ReadFile(path.Join(pd.SrcRoot, pd.ImportPath, fname))
+		src, err := ioutil.ReadFile(path.Join(pd.Dir, fname))
 		if err != nil {
 			return nil, nil, err
 		}
@@ -194,7 +194,7 @@ func parsePackageSource(pkg string) (*types.Scope, *doc.Package, error) {
 	}
 
 	cfg := types.Config{
-		Importer: importer.Default(),
+		Importer: importer.ForCompiler(fset, "source", nil),
 	}
 	info := types.Info{
 		Defs: make(map[*ast.Ident]types.Object),

--- a/assert/assertions.go
+++ b/assert/assertions.go
@@ -22,7 +22,7 @@ import (
 	yaml "gopkg.in/yaml.v2"
 )
 
-//go:generate go run ../_codegen/main.go -output-package=assert -template=assertion_format.go.tmpl
+//go:generate sh -c "cd ../_codegen && go build && cd - && ../_codegen/_codegen -output-package=assert -template=assertion_format.go.tmpl"
 
 // TestingT is an interface wrapper around *testing.T
 type TestingT interface {

--- a/assert/forward_assertions.go
+++ b/assert/forward_assertions.go
@@ -13,4 +13,4 @@ func New(t TestingT) *Assertions {
 	}
 }
 
-//go:generate go run ../_codegen/main.go -output-package=assert -template=assertion_forward.go.tmpl -include-format-funcs
+//go:generate sh -c "cd ../_codegen && go build && cd - && ../_codegen/_codegen -output-package=assert -template=assertion_forward.go.tmpl -include-format-funcs"

--- a/require/forward_requirements.go
+++ b/require/forward_requirements.go
@@ -13,4 +13,4 @@ func New(t TestingT) *Assertions {
 	}
 }
 
-//go:generate go run ../_codegen/main.go -output-package=require -template=require_forward.go.tmpl -include-format-funcs
+//go:generate sh -c "cd ../_codegen && go build && cd - && ../_codegen/_codegen -output-package=require -template=require_forward.go.tmpl -include-format-funcs"

--- a/require/requirements.go
+++ b/require/requirements.go
@@ -26,4 +26,4 @@ type BoolAssertionFunc func(TestingT, bool, ...interface{})
 // for table driven tests.
 type ErrorAssertionFunc func(TestingT, error, ...interface{})
 
-//go:generate go run ../_codegen/main.go -output-package=require -template=require.go.tmpl -include-format-funcs
+//go:generate sh -c "cd ../_codegen && go build && cd - && ../_codegen/_codegen -output-package=require -template=require.go.tmpl -include-format-funcs"

--- a/suite/suite.go
+++ b/suite/suite.go
@@ -84,7 +84,7 @@ func Run(t *testing.T, suite TestingSuite) {
 	defer failOnPanic(t)
 
 	suiteSetupDone := false
-	
+
 	methodFinder := reflect.TypeOf(suite)
 	tests := []testing.InternalTest{}
 	for index := 0; index < methodFinder.NumMethod(); index++ {

--- a/suite/suite_test.go
+++ b/suite/suite_test.go
@@ -350,7 +350,7 @@ func TestRunSuite(t *testing.T) {
 type SuiteSetupSkipTester struct {
 	Suite
 
-	setUp bool
+	setUp    bool
 	toreDown bool
 }
 

